build.sbt

```diff
@@ -184,18 +203,61 @@ val `polynote-server` = project.settings(
 ).dependsOn(`polynote-runtime` % "provided", `polynote-runtime` % "test", `polynote-kernel` % "provided", `polynote-kernel` % "test->test")
 
 val sparkSettings = Seq(
+  resolvers ++= {
+    Seq(MavenRepository(name = "Apache Snapshots", root = "https://repository.apache.org/content/repositories/snapshots"))
+  },
   sparkVersion := {
-    scalaVersion.value match {
-      case ver if ver startsWith "2.11" => "2.1.1"
-      case ver                          => "2.4.7"  // Spark 2.4 is first version to publish for scala 2.12
+    scalaBinaryVersion.value match {
+      case "2.11" => "2.1.1"
+      case "2.12" => "3.1.2"
+      case "2.13" => "3.2.0-SNAPSHOT" // No stable release yet, 3.2.0-SNAPSHOT is earliest available for 2.13
+    }
+  },
+  sparkDistVersion := {
+    sparkVersion.value match {
+      case two if two startsWith "2" => "2.1.1"
+      case three                     => "3.1.2"
     }
   },
   libraryDependencies ++= Seq(
     "org.apache.spark" %% "spark-sql" % sparkVersion.value % "provided",
     "org.apache.spark" %% "spark-repl" % sparkVersion.value % "provided",
     "org.apache.spark" %% "spark-sql" % sparkVersion.value % "test",
     "org.apache.spark" %% "spark-repl" % sparkVersion.value % "test"
-  )
+  ),
+  sparkInstallLocation := {
+    sys.env.get("SPARK_INSTALL_LOCATION")
+      //.orElse(sys.env.get("SPARK_HOME").map(file).map(_.getParent))
+      .getOrElse((file(".").getAbsoluteFile / "target" / "spark").toString)
+  },
+  Test / testOptions += Tests.Setup { () =>
+    import sys.process._
+    val baseDir = file(sparkInstallLocation.value)
+    val distVersion = sparkDistVersion.value
+    val pkgName = s"spark-$distVersion-bin-hadoop2.7"
+    val filename = s"$pkgName.tgz"
+    val distUrl = url(s"https://archive.apache.org/dist/spark/spark-$distVersion/$filename")
+    val destDir = baseDir / pkgName
+    if (!destDir.exists()) {
+      baseDir.mkdirs()
+      val pkgFile = baseDir / filename
+      if (!pkgFile.exists()) {
+        pkgFile.createNewFile()
+        println(s"Downloading $distUrl to $pkgFile...")
+        (distUrl #> (baseDir / filename)).!
+      }
+      println(s"Extracting $pkgFile to $baseDir")
+      Seq("tar", "-zxpf", (baseDir / filename).toString, "-C", baseDir.toString).!
+      Seq("pip3", "install", s"pyspark==$distVersion").!
```

Comment:
Hmm, we ran into problems pip installing pyspark in the past, which is why we have all this stuff https://github.com/polynote/polynote/blob/master/polynote-spark/src/main/scala/polynote/kernel/interpreter/python/PySparkInterpreter.scala#L100

I don't recall if our issues were due to our weird internal environment, but I think we shouldn't need to install pyspark here. If the above doesn't work for some reason it might be better to figure out why the logic above isn't working.

Reply:
The pip install is pretty much just to make sure the correct py4j version is installed. Without that, I was getting problems where pyspark was trying to use py4j methods which didn't exist.

👍 to solving it a better way, though.

Reply:
That stuff is [supposed to work for py4j too](https://github.com/polynote/polynote/blob/master/polynote-spark/src/main/scala/polynote/kernel/interpreter/python/PySparkInterpreter.scala#L105), maybe something's changed in Spark 3+?

Reply:
🤔 honestly I'm not sure why it didn't work. I'll comment this out and see what happens in CI (maybe it was just something weird locally?)

Reply:
Actually, no – I'm pretty sure this was something that only failed in CI (since I added a new commit for it)

Reply:
Was able to remove this and everything worked 🤷 I think I was using the wrong Spark distribution.